File: elixir/allergies/lib/allergies.ex
defmodule Allergies do
  use Bitwise

  @allergens ~w/ eggs peanuts shellfish strawberries tomatoes chocolate pollen cats /

  @doc """
  List the allergies for which the corresponding flag bit is true.
  """
  @spec list(non_neg_integer) :: [String.t()]
  def list(flags) do
    for {allergen, idx} <- Enum.with_index(@allergens),
        (flags &&& 1 <<< idx) != 0,
        do: allergen
  end

  @doc """
  Returns whether the corresponding flag bit in 'flags' is set for the item.
  """
  @spec allergic_to?(non_neg_integer, String.t()) :: boolean
  def allergic_to?(flags, item) do
    ## we could do this, reusing the `list` function:
    #
    # item in list(flags)
    #
    ## but this is slightly more efficient:
    0 !==
      for {allergen, idx} <- Enum.with_index(@allergens),
          item == allergen,
          (flags &&& 1 <<< idx) != 0,
          reduce: 0
      do
        result -> result ||| 1 <<< idx
      end
  end
end

File: typescript/allergies/allergies.ts
class Allergies {
  private readonly allergies: string[] = []

  constructor(score: number) {
    for (const [value, allergen] of Allergies.allergens) {
      if (score & value) {
        this.allergies.push(allergen)
      }
    }
  }

  static allergens: Map<number, string> = new Map([
    [1, 'eggs'],
    [2, 'peanuts'],
    [4, 'shellfish'],
    [8, 'strawberries'],
    [16, 'tomatoes'],
    [32, 'chocolate'],
    [64, 'pollen'],
    [128, 'cats'],
  ])

  list(): string[] { return this.allergies }

  allergicTo(allergen: string): boolean {
    return this.allergies.includes(allergen)
  }
}

export { Allergies }
